README.md
# not-so-fast

A Rust library for data validation.

## Features

- Builder-pattern API for reporting validation errors
- Easy composition of validators
- Derive macro implementing validation traits for structs and enums
- Error display with `jq`-like paths to bad values
- Error serialization reflecting input data structure

## Installation

```bash
cargo add not-so-fast # --features derive serde
```

Available cargo features:

- `derive` - enables `Validate` derive macro, disabled by default
- `serde` - enables `serde::Serialize` implementation for `ValidationNode`, disabled by default

## Usage

```rust
use not_so_fast::{Validate, ValidationNode, ValidationError};

#[derive(Validate)]
struct User {
    #[validate(custom = alpha_only, char_length(max = 30))]
    nick: String,
    #[validate(range(min = 15, max = 100))]
    age: u8,
    #[validate(length(max = 3), items(char_length(max = 50)))]
    cars: Vec<String>,
}

fn alpha_only(s: &str) -> ValidationNode {
    ValidationNode::error_if(
        s.chars().any(|c| !c.is_alphanumeric()),
        || ValidationError::with_code("alpha_only")
    )
}

let user = User {
    nick: "**tom1980**".into(),
    age: 200,
    cars: vec![
        "first".into(),
        "second".into(),
        "third".repeat(11),
        "fourth".into(),
    ],
};

let node = user.validate();
assert!(node.is_err());
assert_eq!(
    vec![
        ".age: range: Number not in range: max=100, min=15, value=200",
        ".cars: length: Invalid length: max=3, value=4",
        ".cars[2]: char_length: Invalid character length: max=50, value=55",
        ".nick: alpha_only",
    ].join("\n"),
    node.to_string()
);
```

## Guides

[not-so-fast/examples/manual.rs](not-so-fast/examples/manual.rs) explains how to write custom validators.

[not-so-fast/examples/derive.rs](not-so-fast/examples/derive.rs) shows how to use Validator derive macro.

## Compared to validator

`not-so-fast` attempts to fix issues I stumbled upon when working with the popular [https://github.com/Keats/validator](validator) crate. APIs of the two libraries are similar, but not compatible. Here are the differences of `not-so-fast`:

- Validator composition - Composing validators in `not-so-fast` is simple, since all values - numbers, strings, objects, lists - report validation errors using the same type - `ValidationNode`.
- Essential validators - `not-so-fast` comes with only the essential validators. You're expected to write custom validators to test data against your domain's rules.
- Enum support - `Validate` derive macro works with structs and enums.
- Duck typing - `Validate` derive macro does not look at field types whatsoever. To validate data inside containers, you give derive macro hints on how to access the data (`some`, `items`, `fields` attributes). In return, `not-so-fast` will work with third-party container types, as long as they have std-like API.

## TODO

- Add `matches` derive validator for testing strings against regular expressions
- Add `is_some`/`required` derive validator for testing options
- Consider moving derive validators from `not-so-fast-derive` to `not-so-fast` for better code reusability

## License

Licensed under either of Apache License, Version 2.0 or MIT license at your option.

Unless you explicitly state otherwise, any contribution intentionally submitted for inclusion in this crate by you, as defined in the Apache-2.0 license, shall be dual licensed as above, without any additional terms or conditions.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.0] - 2023-10-12

### Added

- Support for stateful items and fields validators

## [0.1.0] - 2023-03-11

### Added

- API for reporting errors
- Validator derive macro
- CI

[unreleased]: https://github.com/prk3/not-so-fast/compare/0.2.0...HEAD
[0.2.0]:      https://github.com/prk3/not-so-fast/releases/tag/0.2.0
[0.1.0]:      https://github.com/prk3/not-so-fast/releases/tag/0.1.0


not-so-fast/tests/derive.rs
#![allow(dead_code, unused_variables)]

#[macro_use]
extern crate pretty_assertions;

use not_so_fast::*;

mod derive_tests;

#[test]
fn struct_ref() {
    struct StructRef<'a> {
        name_ref: &'a String,
        int_ref: &'a i32,
        vec_ref: &'a Vec<String>,
        slice: &'a [String],
        array_ref: &'a [String; 3],
        option_ref: &'a Option<String>,
        map_ref: &'a std::collections::HashMap<String, u32>,
    }
}

#[test]
fn empty_enum() {
    #[derive(Validate)]
    #[validate(args(a: u64, b: bool), custom = validate_empty_enum)]
    enum EmptyEnum {}

    fn validate_empty_enum(value: &EmptyEnum) -> ValidationNode {
        ValidationNode::ok()
    }

    // We can't construct EmptyEnum. Let's just check if the code compiles.
}

#[test]
fn enum_different_variants() {
    #[derive(Validate)]
    #[validate(
        args(a: u64, b: bool),
        custom(function = validate_enum, args(b))
    )]
    enum Enum {
        NoFields,
        OneField(#[validate(custom(function = validate_string, args(a)))] String),
        TwoFields(
            #[validate(custom(function = validate_string, args(a)))] String,
            u64,
        ),
        OneNamedField {
            #[validate(custom(function = validate_string, args(a)))]
            first: String,
        },
        TwoNamedFields {
            #[validate(custom(function = validate_string, args(a)))]
            first: String,
            second: u64,
        },
    }

    fn validate_string(value: &str, a: u64) -> ValidationNode {
        ValidationNode::error(ValidationError::with_code("s").and_param("a", a))
    }

    fn validate_enum(value: &Enum, b: bool) -> ValidationNode {
        ValidationNode::error(ValidationError::with_code("e").and_param("b", b))
    }

    assert_eq!(
        ".: e: b=false",
        Enum::NoFields.validate_args((100, false)).to_string()
    );
    assert_eq!(
        ".: e: b=false\n.[0]: s: a=100",
        Enum::OneField("x".into())
            .validate_args((100, false))
            .to_string()
    );
    assert_eq!(
        ".: e: b=false\n.[0]: s: a=100",
        Enum::TwoFields("x".into(), 5)
            .validate_args((100, false))
            .to_string()
    );
    assert_eq!(
        ".: e: b=false\n.first: s: a=100",
        Enum::OneNamedField { first: "x".into() }
            .validate_args((100, false))
            .to_string()
    );
    assert_eq!(
        ".: e: b=false\n.first: s: a=100",
        Enum::TwoNamedFields {
            first: "x".into(),
            second: 5
        }
        .validate_args((100, false))
        .to_string()
    );
}


not-so-fast/tests/manual.rs
#[macro_use]
extern crate pretty_assertions;

use not_so_fast::*;

#[test]
fn user() {
    struct User {
        name: String,           // max 5 bytes long
        age: u8,                // min 3
        pet_names: Vec<String>, // at most 2 items, each at most 10 bytes long
    }

    fn validate_user(user: &User) -> ValidationNode {
        ValidationNode::ok()
            .and_field(
                "name",
                ValidationNode::error_if(user.name.len() > 5, || {
                    ValidationError::with_code("length")
                        .and_message("Illegal string length")
                        .and_param("max", 5)
                        .and_param("value", user.name.len())
                }),
            )
            .and_field(
                "age",
                ValidationNode::error_if(user.age < 3, || {
                    ValidationError::with_code("range")
                        .and_message("Number not in range")
                        .and_param("min", 3)
                        .and_param("value", user.age)
                }),
            )
            .and_field(
                "pet_names",
                ValidationNode::ok()
                    .and_error_if(user.pet_names.len() > 2, || {
                        ValidationError::with_code("length")
                            .and_message("Illegal array length")
                            .and_param("max", 2)
                            .and_param("value", user.pet_names.len())
                    })
                    .and_items(user.pet_names.iter(), |_, item| {
                        ValidationNode::error_if(item.len() > 10, || {
                            ValidationError::with_code("length")
                                .and_message("Illegal string length")
                                .and_param("max", 10)
                                .and_param("value", item.len())
                        })
                    }),
            )
    }

    let user = User {
        name: "easdiuasd&&&&&".into(),
        age: 2,
        pet_names: vec![
            "asdjaiu sdhuyags ydgaysd".into(),
            "aisud 8asydahsbdjabsd".into(),
            "a8 7d8diu h788yhkahsd78".into(),
        ],
    };
    let errors = validate_user(&user);

    assert!(errors.is_err());
    assert_eq!(
        ".age: range: Number not in range: min=3, value=2
.name: length: Illegal string length: max=5, value=14
.pet_names: length: Illegal array length: max=2, value=3
.pet_names[0]: length: Illegal string length: max=10, value=24
.pet_names[1]: length: Illegal string length: max=10, value=21
.pet_names[2]: length: Illegal string length: max=10, value=23",
        errors.to_string()
    );
}

#[test]
fn stateful_item_validation() {
    fn validate_unique_numbers(numbers: &[i32]) -> ValidationNode {
        let mut numbers_seen = std::collections::HashSet::new();

        ValidationNode::items(numbers.iter(), |index, item| {
            let first_occurrence = numbers_seen.insert(item);
            ValidationNode::error_if(!first_occurrence, || {
                ValidationError::with_code("non_unique")
                    .and_message(format!("Number {item} at position {index} is non-unique"))
            })
        })
    }

    assert!(validate_unique_numbers(&[1, 4, 5, 6, 8, 9]).is_ok());
    assert!(validate_unique_numbers(&[1, 2, 3, 2, 4, 5, 6, 7]).is_err());
}


not-so-fast/tests/rendering.rs
#[macro_use]
extern crate pretty_assertions;

use not_so_fast::*;

#[test]
fn simple() {
    let errors = ValidationNode::ok()
        .and_error(
            ValidationError::with_code("one")
                .and_message("Test message one")
                .and_param("param1", "value1"),
        )
        .and_field(
            "field_a",
            ValidationNode::ok()
                .and_error(ValidationError::with_code("two"))
                .and_error(ValidationError::with_code("three")),
        )
        .and_field(
            "field_b",
            ValidationNode::ok()
                .and_item(0, ValidationNode::error(ValidationError::with_code("four")))
                .and_item(1, ValidationNode::error(ValidationError::with_code("five")))
                .and_item(1, ValidationNode::error(ValidationError::with_code("six"))),
        )
        .and_field(
            "field_c_~!@#$%^&*()_+",
            ValidationNode::error(ValidationError::with_code("seven")),
        )
        .and_item(
            0,
            ValidationNode::error(ValidationError::with_code("eight")),
        )
        .and_item(
            1,
            ValidationNode::item(2, ValidationNode::error(ValidationError::with_code("nine"))),
        )
        .and_item(
            2,
            ValidationNode::error(
                ValidationError::with_code("c")
                    .and_param("p01", true)
                    .and_param("p02", 1u8)
                    .and_param("p03", 1u16)
                    .and_param("p04", 1u32)
                    .and_param("p05", 1u64)
                    .and_param("p06", 1u128)
                    .and_param("p07", 1i8)
                    .and_param("p08", 1i16)
                    .and_param("p09", 1i32)
                    .and_param("p10", 1i64)
                    .and_param("p11", 1i128)
                    .and_param("p12", 1usize)
                    .and_param("p13", 1.1f32)
                    .and_param("p14", 1.1f64)
                    .and_param("p15", '\n')
                    .and_param("p16", "one\ntwo")
                    .and_param("p17", String::from("three\nfour"))
                    .and_param("p18", ParamValue::Raw("five\nsix".into())),
            ),
        );

    let errors_json = &serde_json::to_string(&errors).unwrap();
    let errors_text = errors.to_string();

    assert_eq!(
        serde_json::json!({
            "errors": [
                "one: Test message one: param1=\"value1\""
            ],
            "field_a": {
                "errors": [
                    "two",
                    "three"
                ]
            },
            "field_b": {
                "0": {
                    "errors": [
                        "four"
                    ]
                },
                "1": {
                    "errors": [
                        "five",
                        "six"
                    ]
                }
            },
            "field_c_~!@#$%^&*()_+": {
                "errors": [
                    "seven"
                ]
            },
            "0": {
                "errors": [
                    "eight"
                ]
            },
            "1": {
                "2": {
                    "errors": [
                        "nine"
                    ]
                }
            },
            "2": {
                "errors": [
                    "c: p01=true, p02=1, p03=1, p04=1, p05=1, p06=1, p07=1, p08=1, p09=1, p10=1, p11=1, p12=1, p13=1.1, p14=1.1, p15='\\n', p16=\"one\\ntwo\", p17=\"three\\nfour\", p18=five\nsix"
                ]
            }
        }),
        serde_json::from_str::<serde_json::Value>(errors_json).unwrap()
    );

    assert_eq!(vec![
            ".: one: Test message one: param1=\"value1\"",
            ".field_a: two",
            ".field_a: three",
            ".field_b[0]: four",
            ".field_b[1]: five",
            ".field_b[1]: six",
            ".\"field_c_~!@#$%^&*()_+\": seven",
            ".[0]: eight",
            ".[1][2]: nine",
            ".[2]: c: p01=true, p02=1, p03=1, p04=1, p05=1, p06=1, p07=1, p08=1, p09=1, p10=1, p11=1, p12=1, p13=1.1, p14=1.1, p15='\\n', p16=\"one\\ntwo\", p17=\"three\\nfour\", p18=five\nsix"
        ].join("\n"),
        errors_text.to_string()
    );
}


not-so-fast/examples/manual.rs
use not_so_fast::*;

fn main() {
    // - Introduction -

    // This example file shows how to write a validator for any data structure.
    // I recommend you read it from start to end, as we build on what we've
    // learned earlier. You can step thought the code with a debugger, to see
    // errors printed one at a time.

    // - Primitives -

    // We will start simple. Let's write a validator that checks if some age
    // is not greater than 150.

    fn validate_age(age: &u32) -> ValidationNode {
        ValidationNode::error_if(*age > 150, || {
            ValidationError::with_code("range")
                .and_message("Number not in range")
                .and_param("max", 150)
                .and_param("value", *age)
        })
    }

    // `ValidationNode::error_if` accepts two parameters. The first is a bool
    // saying whether the value is valid or not. If it's not, `error_if` method
    // calls the second parameter - function returning Error. Here we pass a
    // closure returning an error with code, message, and two params, though
    // only the code is required.

    // Now let's check if the validator works as expected. `is_ok` method on
    // ValidationNode tells us whether value is valid or not.

    assert!(validate_age(&0).is_ok());
    assert!(validate_age(&35).is_ok());
    assert!(validate_age(&70).is_ok());
    assert!(validate_age(&200).is_err());

    // Check out how validation error is displayed.

    print!("{}\n\n", validate_age(&200));

    // Ok, seems good so far. Now, let's validate username string.

    fn validate_username(username: &str) -> ValidationNode {
        ValidationNode::ok()
            .and_error_if(username.len() < 6, || {
                ValidationError::with_code("byte_length")
                    .and_message("String byte length is not allowed")
                    .and_param("min", 6)
                    .and_param("value", username.len())
            })
            .and_error_if(!username.is_ascii(), || {
                ValidationError::with_code("ascii")
                    .and_message("String contains non-ASCII characters")
            })
    }

    // This time we started with no error (ValidationNode::ok()) and added
    // two checks using `and_error_if`.

    assert!(validate_username("user123").is_ok());
    assert!(validate_username("this IS ok TOO").is_ok());
    assert!(validate_username("short").is_err());
    assert!(validate_username("non-ĄŚĆII name").is_err());

    // One value can trigger many errors. Check out how they are displayed.

    print!("{}\n\n", validate_username("§€¢"));

    // Usually strings are validated as a whole. However, you may want to treat
    // them as a collection of characters and generate an error for specific
    // characters. It's possible with list validation, described in the next
    // chapter.

    // - Lists -

    // not-so-fast allows us to validate lists of items. The following code
    // constructs an error for every float that's not between 0.0 and 1.0,
    // in a 3-elements-long list.

    fn validate_color_rgb(numbers: &[f32; 3]) -> ValidationNode {
        ValidationNode::ok()
            .and_item(
                0,
                ValidationNode::error_if(!(0.0..=1.0).contains(&numbers[0]), || {
                    ValidationError::with_code("range")
                }),
            )
            .and_item(
                1,
                ValidationNode::error_if(!(0.0..=1.0).contains(&numbers[1]), || {
                    ValidationError::with_code("range")
                }),
            )
            .and_item(
                2,
                ValidationNode::error_if(!(0.0..=1.0).contains(&numbers[2]), || {
                    ValidationError::with_code("range")
                }),
            )
    }

    assert!(validate_color_rgb(&[0.0, 0.5, 1.0]).is_ok());
    assert!(validate_color_rgb(&[0.0, 5.0, -1.0]).is_err());

    // `add_item` works well if the list has a fixed number of elements.
    // Lists of variable length can be validated using `item` and `and_items`
    // methods, as shown below.

    fn validate_username_list(usernames: &[String]) -> ValidationNode {
        ValidationNode::items(usernames.iter(), |_index, username| {
            validate_username(username)
        })
    }

    assert!(validate_username_list(&["ann001".into(), "bob002".into()]).is_ok());
    assert!(validate_username_list(&["short".into(), "≠»²«³¢".into()]).is_err());

    // Notice how we've reused username validation function.

    // To validate both the length of the list and all items in it, use code
    // like this.

    fn validate_username_list_2(usernames: &[String]) -> ValidationNode {
        ValidationNode::ok()
            .and_error_if(usernames.len() > 2, || ValidationError::with_code("length"))
            .and_items(usernames.iter(), |_index, username| {
                validate_username(username)
            })
    }

    assert!(validate_username_list_2(&["ann001".into(), "bob002".into()]).is_ok());
    assert!(
        validate_username_list_2(&["ann001".into(), "bob002".into(), "chris003".into()]).is_err()
    );

    // Nice. Let's see how it's displayed.

    print!(
        "{}\n\n",
        validate_username_list_2(&["ok_username".into(), "short".into(), "»»»bad«««".into(),])
    );

    // - Objects -

    // What is we wanted to validate an entire struct? Methods `field` and
    // `and_field` allow us to do just that.

    #[derive(Clone)]
    struct User {
        username: String,
        age: u32,
        favorite_color_rgb: [f32; 3],
        friends: Vec<String>,
    }

    fn validate_user(user: &User) -> ValidationNode {
        ValidationNode::ok()
            .and_field("username", validate_username(&user.username))
            .and_field("age", validate_age(&user.age))
            .and_field(
                "favorite_color_rgb",
                validate_color_rgb(&user.favorite_color_rgb),
            )
            .and_field("friends", validate_username_list_2(&user.friends))
    }

    // Because we've defined validators of user properties before, validator of
    // User struct is trivial. Now let's see it in action.

    let ok_user = User {
        username: "hello_world".into(),
        age: 1,
        favorite_color_rgb: [0.0, 1.0, 0.0],
        friends: vec!["foo_bar".into()],
    };

    let bad_user = User {
        username: "¬.¬".into(),
        age: 2000,
        favorite_color_rgb: [-0.5, 5.0, f32::INFINITY],
        friends: vec!["short".into(), "€".into(), "third³".into()],
    };

    assert!(validate_user(&ok_user).is_ok());
    assert!(validate_user(&bad_user).is_err());

    print!("{}\n\n", validate_user(&bad_user));

    // Wonderful. What if an object must uphold some invariant, meaning its
    // properties depend on each other? We can add an error at the object level.

    fn validate_mean_user(user: &User) -> ValidationNode {
        validate_user(user).and_error_if(
            user.username.starts_with("mean") && !user.friends.is_empty(),
            || {
                ValidationError::with_code("mean_user_invariant")
                    .and_message("Mean users can't have friends")
            },
        )
    }

    let mean_user = User {
        username: "mean dan".into(),
        age: 30,
        favorite_color_rgb: [0.0, 0.0, 0.0],
        friends: vec!["nice ethan".into()],
    };

    assert!(validate_mean_user(&mean_user).is_err());

    print!("{}\n\n", validate_mean_user(&mean_user));

    // The last element of the validation puzzle is checking dynamic objects
    // (also called maps, dictionaries or associative arrays). They can be
    // validated using `fields` and `and_fields` methods.

    fn validate_product_prices(
        product_prices: &std::collections::HashMap<String, f32>,
    ) -> ValidationNode {
        ValidationNode::fields(product_prices.iter(), |_key, value| {
            ValidationNode::error_if(*value < 0.0, || ValidationError::with_code("range"))
        })
    }

    let mut ok_product_prices = std::collections::HashMap::new();
    ok_product_prices.insert("pizza".into(), 10.0);
    ok_product_prices.insert("apple".into(), 1.5);
    ok_product_prices.insert("chocolate".into(), 3.0);

    let mut bad_product_prices = ok_product_prices.clone();
    bad_product_prices.insert("aether".into(), -10.0);

    assert!(validate_product_prices(&bad_product_prices).is_err());

    print!("{}\n\n", validate_product_prices(&bad_product_prices));

    // - Keys -

    // Functions you pass to `items`, `and_items`, `fields`, and `and_fields`
    // methods must accept two parameters: the first is index/key and the
    // second is item/value. It's done like this because we sometimes want
    // to check indexes/keys as well. Here is an example.

    #[derive(Clone)]
    struct OrderedProduct {
        id: String,
        #[allow(unused)]
        quantity: u32,
    }

    fn validate_order(order: &std::collections::HashMap<String, OrderedProduct>) -> ValidationNode {
        ValidationNode::fields(order.iter(), |id, product| {
            ValidationNode::error_if(id != &product.id, || {
                ValidationError::with_code("product_id_invariant")
                    .and_message("Product ID key in order object does not match product's ID")
            })
        })
    }

    // This validator makes sure that product IDs used as keys in order object
    // match with product IDs held in product objects (order object values).

    let mut ok_order = std::collections::HashMap::new();
    ok_order.insert(
        "mug-23".into(),
        OrderedProduct {
            id: "mug-23".into(),
            quantity: 2,
        },
    );

    assert!(validate_order(&ok_order).is_ok());

    let mut bad_order = ok_order.clone();
    bad_order.insert(
        "pencil-3".into(),
        OrderedProduct {
            id: "bucket-39".into(),
            quantity: 1,
        },
    );

    assert!(validate_order(&bad_order).is_err());

    // Notice that, despite accepting two parameters, the error function above
    // can return only one error object. If both index/key and item/value can
    // produce a validation error, it's our responsibility to communicate the
    // source of an error clearly.

    // We could report errors on the object that has invalid keys.

    fn validate_product_prices_alt_1(
        product_prices: &std::collections::HashMap<String, f32>,
    ) -> ValidationNode {
        ValidationNode::ok()
            .and_errors(
                product_prices.keys().filter_map(|name| {
                    (!name.is_ascii()).then(|| ValidationError::with_code("ascii"))
                }),
            )
            .and_fields(product_prices.iter(), |_name, price| {
                ValidationNode::error_if(price.is_infinite(), || {
                    ValidationError::with_code("infinite_price")
                })
            })
    }

    // Or mix add "key" errors to "value" error object.

    fn validate_product_prices_alt_2(
        product_prices: &std::collections::HashMap<String, f32>,
    ) -> ValidationNode {
        ValidationNode::fields(product_prices.iter(), |name, price| {
            ValidationNode::ok()
                .and_error_if(!name.is_ascii(), || ValidationError::with_code("ascii"))
                .and_error_if(price.is_infinite(), || {
                    ValidationError::with_code("infinite_price")
                })
        })
    }

    // Or introduce a fictional object with "key" and "value" fields.

    fn validate_product_prices_alt_3(
        product_prices: &std::collections::HashMap<String, f32>,
    ) -> ValidationNode {
        ValidationNode::fields(product_prices.iter(), |name, price| {
            ValidationNode::ok()
                .and_field(
                    "key",
                    ValidationNode::error_if(!name.is_ascii(), || {
                        ValidationError::with_code("ascii")
                    }),
                )
                .and_field(
                    "value",
                    ValidationNode::error_if(price.is_infinite(), || {
                        ValidationError::with_code("infinite_price")
                    }),
                )
        })
    }

    // There are many possibilities, each producing a different error message.
    // You have to choose strategy that makes the most sense for you.

    let mut terrible_product_prices = std::collections::HashMap::new();
    terrible_product_prices.insert("½".into(), f32::INFINITY);

    print!(
        "{}\n\n",
        validate_product_prices_alt_1(&terrible_product_prices)
    );
    print!(
        "{}\n\n",
        validate_product_prices_alt_2(&terrible_product_prices)
    );
    print!(
        "{}\n\n",
        validate_product_prices_alt_3(&terrible_product_prices)
    );

    // - Parametrization -

    // Since manually written validators are regular, unconstrained functions,
    // there is nothing preventing us from parameterizing them.

    struct Profile {
        name: String,
        bio: String,
        images: Vec<String>,
    }

    fn validate_bio(bio: &str, max_char_length: usize) -> ValidationNode {
        let len = bio.chars().count();
        ValidationNode::error_if(len > max_char_length, || {
            ValidationError::with_code("length")
                .and_message("Illegal string length")
                .and_param("max", max_char_length)
                .and_param("value", len)
        })
    }

    fn validate_images(images: &[String], multi_image: bool) -> ValidationNode {
        let limit = if multi_image { 10 } else { 1 };
        ValidationNode::error_if(images.len() > limit, || {
            ValidationError::with_code("length")
                .and_message("Illegal list length")
                .and_param("max", limit)
                .and_param("value", images.len())
        })
    }

    fn validate_profile(
        profile: &Profile,
        max_bio_char_length: usize,
        multi_image: bool,
    ) -> ValidationNode {
        ValidationNode::ok()
            .and_field("name", validate_username(&profile.name))
            .and_field("bio", validate_bio(&profile.bio, max_bio_char_length))
            .and_field("images", validate_images(&profile.images, multi_image))
    }

    let profile = Profile {
        name: "foo_bar_3".into(),
        bio: "x".repeat(1000),
        images: vec!["one.jpg".into(), "two.jpg".into(), "three.jpg".into()],
    };

    assert!(validate_profile(&profile, 2000, true).is_ok());
    assert!(validate_profile(&profile, 2000, false).is_err());
    assert!(validate_profile(&profile, 300, true).is_err());
    assert!(validate_profile(&profile, 300, false).is_err());

    // To make your validators compatible with Validator derive macro, place
    // validator parameters after the reference to value.

    // - Merging -

    // Sometimes it's convenient to split validation logic into two or more
    // functions. In case of comments, one validator could check the length
    // of the comment text, while the other could look for illegal words. To
    // merge two errors together, use `merge` method.

    struct Comment {
        author: String,
        text: String,
    }

    fn validate_text_length(text: &str) -> ValidationNode {
        ValidationNode::error_if(text.len() > 500, || {
            ValidationError::with_code("byte_length").and_message("Illegal string byte length")
        })
    }

    fn validate_text_content(text: &str) -> ValidationNode {
        let illegal_words = ["pineapple", "bash", "truck"];
        let contains_illegal_word = illegal_words.iter().any(|word| text.contains(word));

        ValidationNode::error_if(contains_illegal_word, || {
            ValidationError::with_code("illegal_word").and_message("Text contains illegal word")
        })
    }

    fn validate_comment(comment: &Comment) -> ValidationNode {
        ValidationNode::ok()
            .and_field("author", validate_username(&comment.author))
            .and_field(
                "text",
                ValidationNode::ok()
                    .merge(validate_text_length(&comment.text))
                    .merge(validate_text_content(&comment.text)),
            )
    }

    let bad_comment_length = Comment {
        author: "ok".into(),
        text: "x".repeat(1000),
    };

    let bad_comment_content = Comment {
        author: "ok".into(),
        text: "I love pineapple pizza".into(),
    };

    assert!(validate_comment(&bad_comment_length).is_err());
    assert!(validate_comment(&bad_comment_content).is_err());

    // `merge` method not only moves all value errors from the error argument
    // to `self`, but also recursively combines field/item errors.

    // - Traits -

    // Until now we've been writing stand-alone validation functions. To
    // associate a validation function with the validated type, implement
    // `ValidateArgs` for that type.

    impl<'arg> ValidateArgs<'arg> for User {
        type Args = ();

        fn validate_args(&self, (): Self::Args) -> ValidationNode {
            validate_user(self)
        }
    }

    // Associated type `Args` is a tuple with validation parameters. In the
    // example above the tuple has zero elements, since user validation
    // does not expect any arguments.

    // `Profile` does have validation parameters, so the implementation of
    // `ValidateArgs` trait would look like this.

    impl<'arg> ValidateArgs<'arg> for Profile {
        type Args = (usize, bool);

        fn validate_args(&self, (max_bio_len, multi_image): Self::Args) -> ValidationNode {
            validate_profile(self, max_bio_len, multi_image)
        }
    }

    // The generic lifetime 'arg can be used to pass expensive-to-copy argument
    // to the validator by reference.

    struct Shape(String);

    impl<'arg> ValidateArgs<'arg> for Shape {
        type Args = (&'arg [&'arg str],);

        fn validate_args(&self, (legal_shapes,): Self::Args) -> ValidationNode {
            ValidationNode::error_if(!legal_shapes.contains(&self.0.as_ref()), || {
                ValidationError::with_code("illegal_shape")
            })
        }
    }

    // To validate data through `ValidateArgs` implementation, just call
    // `validate_args` method providing necessary arguments.

    const LEGAL_SHAPES_2D: [&str; 3] = ["triangle", "square", "circle"];
    const LEGAL_SHAPES_3D: [&str; 3] = ["sphere", "magenta", "yellow"];

    assert!(Shape("square".into())
        .validate_args((&LEGAL_SHAPES_2D[..],))
        .is_ok());

    assert!(Shape("square".into())
        .validate_args((&LEGAL_SHAPES_3D[..],))
        .is_err());

    assert!(ok_user.validate_args(()).is_ok());

    // Because validators usually aren't parameterized, types implementing
    // `ValidateArgs<Args=()>` automatically implement simpler `Validate` trait.
    // Using `Validate` in trait bounds is preferred, as it shortens your code.

    fn clone_and_validate_1<'a, T: Clone + ValidateArgs<'a, Args = ()>>(
        value: &T,
    ) -> Result<T, ValidationNode> {
        let clone: T = value.clone();
        clone.validate_args(()).result()?;
        Ok(clone)
    }

    assert!(clone_and_validate_1(&ok_user).is_ok());

    fn clone_and_validate_2<T: Clone + Validate>(value: &T) -> Result<T, ValidationNode> {
        let clone: T = value.clone();
        clone.validate().result()?;
        Ok(clone)
    }

    assert!(clone_and_validate_2(&bad_user).is_err());
}
